/**
 * \file draw3dParametersControlWidget.cpp
 * \attention This file is automatically generated and should not be in general modified manually
 *
 * \date MMM DD, 20YY
 * \author autoGenerator
 */

#include "draw3dParametersControlWidget.h"
#include "ui_draw3dParametersControlWidget.h"
#include <memory>
#include "qSettingsGetter.h"
#include "qSettingsSetter.h"

#include "rgbColorParametersControlWidget.h"
#include "rgbColorParametersControlWidget.h"
#include "rgbColorParametersControlWidget.h"
#include "rgbColorParametersControlWidget.h"

Draw3dParametersControlWidget::Draw3dParametersControlWidget(QWidget *parent, bool _autoInit, QString _rootPath)
    : ParametersControlWidgetBase(parent)
    , mUi(new Ui::Draw3dParametersControlWidget)
    , autoInit(_autoInit)
    , rootPath(_rootPath)
{
    mUi->setupUi(this);

    QObject::connect(mUi->styleComboBox, SIGNAL(currentIndexChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->pointColorControlWidget, SIGNAL(paramsChanged()), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->pointColorOverrideCheckBox, SIGNAL(stateChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->pointSizeSpinBox, SIGNAL(valueChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->edgeColorControlWidget, SIGNAL(paramsChanged()), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->edgeColorOverrideCheckBox, SIGNAL(stateChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->edgeWidthSpinBox, SIGNAL(valueChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->faceColorControlWidget, SIGNAL(paramsChanged()), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->faceColorOverrideCheckBox, SIGNAL(stateChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->showCaptionCheckBox, SIGNAL(stateChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->fontSizeSpinBox, SIGNAL(valueChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->fontWidthSpinBox, SIGNAL(valueChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->fontColorControlWidget, SIGNAL(paramsChanged()), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->textureCorrodinatesComboBox, SIGNAL(currentIndexChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->textureAlphaSpinBox, SIGNAL(valueChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->textureScaleSpinBox, SIGNAL(valueChanged(double)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->decalMatrixTypeSpinBox, SIGNAL(valueChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->decalLeftCamCheckBox, SIGNAL(stateChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->decalLeftAlphaSpinBox, SIGNAL(valueChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->decalRightCamCheckBox, SIGNAL(stateChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->decalRightAlphaSpinBox, SIGNAL(valueChanged(int)), this, SIGNAL(paramsChanged()));
}

Draw3dParametersControlWidget::~Draw3dParametersControlWidget()
{

    delete mUi;
}

void Draw3dParametersControlWidget::loadParamWidget(WidgetLoader &loader)
{
    std::unique_ptr<Draw3dParameters> params(createParameters());
    loader.loadParameters(*params, rootPath);
    setParameters(*params);
}

void Draw3dParametersControlWidget::saveParamWidget(WidgetSaver  &saver)
{
    saver.saveParameters(*std::unique_ptr<Draw3dParameters>(createParameters()), rootPath);
}

void Draw3dParametersControlWidget::getParameters(Draw3dParameters& params) const
{
    params = *std::unique_ptr<Draw3dParameters>(createParameters());
}


Draw3dParameters *Draw3dParametersControlWidget::createParameters() const
{

    /**
     * We should think of returning parameters by value or saving them in a preallocated place
     **/


    return new Draw3dParameters(
          static_cast<Draw3dStyle::Draw3dStyle>(mUi->styleComboBox->currentIndex())
        , *std::unique_ptr<RgbColorParameters>(mUi->pointColorControlWidget->createParameters())
        , mUi->pointColorOverrideCheckBox->isChecked()
        , mUi->pointSizeSpinBox->value()
        , *std::unique_ptr<RgbColorParameters>(mUi->edgeColorControlWidget->createParameters())
        , mUi->edgeColorOverrideCheckBox->isChecked()
        , mUi->edgeWidthSpinBox->value()
        , *std::unique_ptr<RgbColorParameters>(mUi->faceColorControlWidget->createParameters())
        , mUi->faceColorOverrideCheckBox->isChecked()
        , mUi->showCaptionCheckBox->isChecked()
        , mUi->fontSizeSpinBox->value()
        , mUi->fontWidthSpinBox->value()
        , *std::unique_ptr<RgbColorParameters>(mUi->fontColorControlWidget->createParameters())
        , static_cast<Draw3dTextureGen::Draw3dTextureGen>(mUi->textureCorrodinatesComboBox->currentIndex())
        , mUi->textureAlphaSpinBox->value()
        , mUi->textureScaleSpinBox->value()
        , mUi->decalMatrixTypeSpinBox->value()
        , mUi->decalLeftCamCheckBox->isChecked()
        , mUi->decalLeftAlphaSpinBox->value()
        , mUi->decalRightCamCheckBox->isChecked()
        , mUi->decalRightAlphaSpinBox->value()
    );
}

void Draw3dParametersControlWidget::setParameters(const Draw3dParameters &input)
{
    // Block signals to send them all at once
    bool wasBlocked = blockSignals(true);
    mUi->styleComboBox->setCurrentIndex(input.style());
    mUi->pointColorControlWidget->setParameters(input.pointColor());
    mUi->pointColorOverrideCheckBox->setChecked(input.pointColorOverride());
    mUi->pointSizeSpinBox->setValue(input.pointSize());
    mUi->edgeColorControlWidget->setParameters(input.edgeColor());
    mUi->edgeColorOverrideCheckBox->setChecked(input.edgeColorOverride());
    mUi->edgeWidthSpinBox->setValue(input.edgeWidth());
    mUi->faceColorControlWidget->setParameters(input.faceColor());
    mUi->faceColorOverrideCheckBox->setChecked(input.faceColorOverride());
    mUi->showCaptionCheckBox->setChecked(input.showCaption());
    mUi->fontSizeSpinBox->setValue(input.fontSize());
    mUi->fontWidthSpinBox->setValue(input.fontWidth());
    mUi->fontColorControlWidget->setParameters(input.fontColor());
    mUi->textureCorrodinatesComboBox->setCurrentIndex(input.textureCorrodinates());
    mUi->textureAlphaSpinBox->setValue(input.textureAlpha());
    mUi->textureScaleSpinBox->setValue(input.textureScale());
    mUi->decalMatrixTypeSpinBox->setValue(input.decalMatrixType());
    mUi->decalLeftCamCheckBox->setChecked(input.decalLeftCam());
    mUi->decalLeftAlphaSpinBox->setValue(input.decalLeftAlpha());
    mUi->decalRightCamCheckBox->setChecked(input.decalRightCam());
    mUi->decalRightAlphaSpinBox->setValue(input.decalRightAlpha());
    blockSignals(wasBlocked);
    emit paramsChanged();
}

void Draw3dParametersControlWidget::setParametersVirtual(void *input)
{
    // Modify widget parameters from outside
    Draw3dParameters *inputCasted = static_cast<Draw3dParameters *>(input);
    setParameters(*inputCasted);
}
